New Module: CSS Selectors (#28062)
mdn/content: files/en-us/web/css/css_selectors/index.md

- {{CSSXref(":default")}}
- {{CSSXref(":defined")}}
- {{CSSXref(":dir", ":dir()")}}
- {{CSSXref(":disabled")}}
- {{CSSXref(":empty")}}
- {{CSSXref(":enabled")}}
- {{CSSXref(":first-child")}}
- {{CSSXref(":first-of-type")}}
- {{CSSXref(":focus")}}
- {{CSSXref(":focus-visible")}}
- {{CSSXref(":focus-within")}}
- {{CSSXref(":fullscreen")}}
- {{CSSXref(":future")}}
- {{CSSXref(":has", ":has()")}}
- {{CSSXref(":hover")}}
- {{CSSXref(":indeterminate")}}
- {{CSSXref(":in-range")}}
- {{CSSXref(":invalid")}}
- {{CSSXref(":is", ":is()")}}
- {{CSSXref(":lang", ":lang()")}}
- {{CSSXref(":last-child")}}
- {{CSSXref(":last-of-type")}}
- {{CSSXref(":link")}}
- {{CSSXref(":local-link")}}
- `:matches()` (obsolete legacy selector alias for {{CSSXref( ":is", ":is()")}})
- {{CSSXref(":modal")}}
- {{CSSXref(":muted")}}
- {{CSSXref(":not", ":not()")}}
- {{CSSXref(":nth-child", ":nth-child()")}}
- {{CSSXref(":nth-col", ":nth-col()")}}
- {{CSSXref(":nth-last-child", ":nth-last-child()")}}
- {{CSSXref(":nth-last-col", ":nth-last-col()")}}
- {{CSSXref(":nth-last-of-type", ":nth-last-of-type()")}}
- {{CSSXref(":nth-of-type", ":nth-of-type()")}}
- {{CSSXref(":only-child")}}
- {{CSSXref(":only-of-type")}}

> - {{CSSXref(":open")}}

Author: ```suggestion
```

- [Using the `:target` pseudo-class in selectors](/en-US/docs/Web/CSS/CSS_selectors/Using_the_:target_pseudo-class_in_selectors)

  - : Learn how to use the {{CSSXref(":target")}} pseudo-class to style the target element a URL's fragment identifier.

- [Learn: UI pseudo-classes](/en-US/docs/Learn/Forms/UI_pseudo-classes)

  - : Learn the different UI pseudo-classes available for styling forms in different states.

- [Locating DOM elements using selectors](/en-US/docs/Web/API/Document_object_model/Locating_DOM_elements_using_selectors)

  - : The selectors API enables using selectors in JavaScript to retrieve element nodes from the DOM.

> ## Related concepts

Author: ```suggestion
## Related concepts

- {{CSSXref(":popover-open")}} pseudo-class
```

---
title: CSS selectors
slug: Web/CSS/CSS_selectors
page-type: css-module
spec-urls: https://drafts.csswg.org/selectors/
---

{{CSSRef}}

The **CSS selectors** module defines the patterns to select elements to which a set of CSS rules are then applied along with their {{cssxref("specificity")}}. The CSS selectors module provides us with more than 60 selectors and five combinators. Other modules provide additional pseudo-class selectors and pseudo-elements.

In CSS, selectors are patterns used to match, or select, the elements you want to style. Selectors are also used in JavaScript to enable selecting which DOM nodes to return as a [`NodeList`](/en-US/docs/Web/API/NodeList).

Selectors, whether used in CSS or JS, enable targeting HTML elements based on their type, attributes, current states, and even position in the DOM. Combinators enable being more precise in the elements you select by enabling specifying elements based on their relationship to other elements.

## Reference

> ### Combinators and seperators

Author: namespace is in a different spec. That's what makes this quite the challenge

- {{CSSXref(":current", ":current()")}}
- {{CSSXref(":default")}}
- {{CSSXref(":defined")}}
- {{CSSXref(":dir", ":dir()")}}
- {{CSSXref(":disabled")}}
- {{CSSXref(":empty")}}
- {{CSSXref(":enabled")}}
- {{CSSXref(":first-child")}}
- {{CSSXref(":first-of-type")}}
- {{CSSXref(":focus")}}
- {{CSSXref(":focus-visible")}}
- {{CSSXref(":focus-within")}}
- {{CSSXref(":fullscreen")}}
- {{CSSXref(":future")}}
- {{CSSXref(":has", ":has()")}}
- {{CSSXref(":hover")}}
- {{CSSXref(":indeterminate")}}
- {{CSSXref(":in-range")}}
- {{CSSXref(":invalid")}}
- {{CSSXref(":is", ":is()")}}
- {{CSSXref(":lang", ":lang()")}}
- {{CSSXref(":last-child")}}
- {{CSSXref(":last-of-type")}}
- {{CSSXref(":link")}}
- {{CSSXref(":local-link")}}
- {{CSSXref( ":is", ":matches()")}}
- {{CSSXref(":modal")}}
- {{CSSXref(":muted")}}
- {{CSSXref(":not", ":not()")}}
- {{CSSXref(":nth-child", ":nth-child()")}}
- {{CSSXref(":nth-col", ":nth-col()")}}
- {{CSSXref(":nth-last-child", ":nth-last-child()")}}
- {{CSSXref(":nth-last-col", ":nth-last-col()")}}
- {{CSSXref(":nth-last-of-type", ":nth-last-of-type()")}}
- {{CSSXref(":nth-of-type", ":nth-of-type()")}}
- {{CSSXref(":only-child")}}
- {{CSSXref(":only-of-type")}}

> - {{CSSXref(":open")}}

Author: Thank you for that info. I'll add it to the related. It's currently in the HTML spec, not yet in the CSS spec. Will add it and change the spec on the popover-open page when it is in the spec.

- {{CSSXref(":optional")}}
- {{CSSXref(":out-of-range")}}
- {{CSSXref(":past")}}
- {{CSSXref(":paused")}}
- {{CSSXref(":picture-in-picture")}}
- {{CSSXref(":placeholder-shown")}}
- {{CSSXref(":playing")}}
- {{CSSXref(":read-only")}}
- {{CSSXref(":read-write")}}
- {{CSSXref(":required")}}
- {{CSSXref(":root")}}
- {{CSSXref(":scope")}}
- {{CSSXref(":seeking")}}
- {{CSSXref(":stalled")}}
- {{CSSXref(":target")}}
- {{CSSXref(":target-within")}}
- {{CSSXref(":user-invalid")}}
- {{CSSXref(":user-valid")}}
- {{CSSXref(":valid")}}
- {{CSSXref(":visited")}}
- {{CSSXref(":volume-locked")}}
- {{CSSXref(":where", ":where()")}}
- [`:-webkit-` pseudo-classes](/en-US/docs/Web/CSS/WebKit_Extensions#pseudo-classes)

> - [Attribute selectors](/en-US/docs/Web/CSS/Attribute_selectors)
> - [Class selector](/en-US/docs/Web/CSS/Class_selectors)

Reviewer: ```suggestion
### Selectors 

- [Attribute selectors](/en-US/docs/Web/CSS/Attribute_selectors)
- [Class selectors](/en-US/docs/Web/CSS/Class_selectors)
```

> - [ID selector](/en-US/docs/Web/CSS/ID_selectors)
> - [Type selector](/en-US/docs/Web/CSS/Type_selectors)
> - [Universal selector](/en-US/docs/Web/CSS/Universal_selectors)

Reviewer: ```suggestion
- [ID selectors](/en-US/docs/Web/CSS/ID_selectors)
- [Type selectors](/en-US/docs/Web/CSS/Type_selectors)
- [Universal selectors](/en-US/docs/Web/CSS/Universal_selectors)
```

> ## Definitions

Reviewer: Can we instead call this section "Terms" instead because what we're listing is terms and linking to definitions?

```suggestion
## Terms
```

- {{glossary("Pseudo-class")}} glossary term

> - [Functional pseudo class](/en-US/docs/Web/CSS/Pseudo-classes#functional_pseudo-classes)
> - [Combinator](/en-US/docs/Web/CSS/CSS_selectors/Selectors_and_combinators#combinators)

Reviewer: ```suggestion
- [Functional pseudo-classes](/en-US/docs/Web/CSS/Pseudo-classes#functional_pseudo-classes)
- [Combinators](/en-US/docs/Web/CSS/CSS_selectors/Selectors_and_combinators#combinators)
```

- [Simple selector](/en-US/docs/Web/CSS/CSS_selectors/Selectors_and_combinators#simple-selector)
- [Compound selector](/en-US/docs/Web/CSS/CSS_selectors/Selectors_and_combinators#compound-selector)
- [Complex selector](/en-US/docs/Web/CSS/CSS_selectors/Selectors_and_combinators#complex-selector)
- [Relative selector](/en-US/docs/Web/CSS/CSS_selectors/Selectors_and_combinators#relative-selector)
- [Selector list](/en-US/docs/Web/CSS/Selector_list)

> - {{CSSXref("Specificity")}}

Reviewer: the macro is adding code styling on the word, so replacing the link with the slug
```suggestion
- [Specificity](/en-US/docs/Web/CSS/Specificity)
```

## Guides

- [CSS selectors and combinators](/en-US/docs/Web/CSS/CSS_selectors/Selectors_and_combinators)

>   - : Overview of the different types of simple selectors and the various combinators defined in the CSS selectors and pseudo modules.

Reviewer: ```suggestion
  - : Overview of the different types of simple selectors and various combinators defined in the CSS selectors and the CSS pseudo modules.
```

- [CSS selector structure](/en-US/docs/Web/CSS/CSS_selectors/Selector_structure)

>   - : The structure of CSS selectors and the terminology introduced in the selectors module, from "simple selector" to "forgiving relative selector list".

Reviewer: ```suggestion
  - : Explanation of the structure of CSS selectors and the terminologies introduced in the CSS selectors module, ranging from "simple selector" to "forgiving relative selector list".
```

> - [Attribute selectors](/en-US/docs/Web/CSS/Attribute_selectors)

Reviewer: Should we have this here, it is not really a guide? We have it covered in the previous list. Also we are not listing other selector types here in the Guides section

  - : Selects elements that have the given attribute, with optional substring values.

> - [Pseudo classes](/en-US/docs/Web/CSS/Pseudo-classes)

Reviewer: Same comment - do we need to list it here? We have it covered in the previous list

Author: kept pseudo and expanded the explaination

  - : Pseudo-classes allow the selection of elements based on state information that is not contained in the document tree.

- [Learn: CSS selectors](/en-US/docs/Learn/CSS/Building_blocks/Selectors)

  - : Part of CSS building blocks, includes tutorials on [Type, class, and ID selectors](/en-US/docs/Learn/CSS/Building_blocks/Selectors/Type_Class_and_ID_Selectors), [Attribute selectors](/en-US/docs/Learn/CSS/Building_blocks/Selectors/Attribute_selectors), [Pseudo-classes and pseudo-elements](/en-US/docs/Learn/CSS/Building_blocks/Selectors/Pseudo-classes_and_pseudo-elements), [Combinators](/en-US/docs/Learn/CSS/Building_blocks/Selectors/Combinators), [Cascade, specificity, and inheritance](/en-US/docs/Learn/CSS/Building_blocks/Cascade_and_inheritance), and [Cascade layers](/en-US/docs/Learn/CSS/Building_blocks/Cascade_layers).

> - [Using the `:target` pseudo-class in selectors](/en-US/docs/Web/CSS/CSS_selectors/Using_the_:target_pseudo-class_in_selectors)

Author: if this wasn't so unrelated it would make sense, but I am unclear why target is where it is.

## Related concepts

- CSS scoping module

  - {{CSSXref(":host")}} pseudo-class
  - {{CSSXref(":host_function", ":host()")}} pseudo-class
  - {{cssxref(":host-context", ":host-context()")}} pseudo-class
  - {{CSSXref("::slotted")}} pseudo-element

- [CSS pseudo-element module](/en-US/docs/Web/CSS/CSS_pseudo) (representing entities not included in HTML)

  - {{CSSXref("::after")}}
  - {{CSSXref("::before")}}
  - {{CSSXref("::file-selector-button")}}
  - {{CSSXref("::first-letter")}}
  - {{CSSXref("::first-line")}}
  - {{CSSXref("::grammar-error")}}
  - {{CSSXref("::marker")}}
  - {{CSSXref("::placeholder")}}
  - {{CSSXref("::selection")}}
  - {{CSSXref("::spelling-error")}}
  - {{CSSXref("::target-text")}}

- [CSS shadow parts module](/en-US/docs/Web/CSS/CSS_shadow_parts)

  - {{CSSXref("::part")}} pseudo-element

- [CSS positioned layout module](/en-US/docs/Web/CSS/CSS_positioned_layout)

  - {{CSSxRef("::backdrop")}}

- Other [pseudo-elements](/en-US/docs/Web/CSS/Pseudo-elements)

  - {{CSSxRef("::cue")}}
  - {{CSSxRef("::cue-region")}}

- {{CSSXref("@namespace")}} at-rule

- {{cssxref("important", "!important")}}

> - {{cssxref("Specificity")}}
> - [The cascade](/en-US/docs/Web/CSS/Cascade)

Reviewer: ```suggestion
- [Specificity](/en-US/docs/Web/CSS/Specificity)
- [Cascade](/en-US/docs/Web/CSS/Cascade)
```

- {{domxref("Document.querySelector")}} method
- {{domxref("Document.querySelectorAll")}} method
- {{domxref("NodeList.forEach()")}} method

## Specifications

{{Specifications}}

## See also

- [CSS pseudo-element module](/en-US/docs/Web/CSS/CSS_pseudo)
- [CSS cascade and inheritance module](/en-US/docs/Web/CSS/CSS_cascade)

> - [Cascade origin and importance](/en-US/docs/Web/CSS/Cascade)

Reviewer: we are covering this under Related concepts as well
```suggestion

```
